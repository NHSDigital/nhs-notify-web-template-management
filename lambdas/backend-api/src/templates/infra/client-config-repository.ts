import { z } from 'zod';
import NodeCache from 'node-cache';
import { GetParameterCommand, SSMClient } from '@aws-sdk/client-ssm';
import { Client, $Client } from 'nhs-notify-backend-client';
import { Logger } from 'nhs-notify-web-template-management-utils/logger';
import { parseJsonPreprocessor } from '@backend-api/utils/zod-json-preprocessor';

const $ClientProcessed = z.preprocess(parseJsonPreprocessor, $Client);

export class ClientConfigRepository {
  constructor(
    private readonly ssmKeyPrefix: string,
    private readonly ssmClient: SSMClient,
    private readonly cache: NodeCache,
    private readonly logger: Logger
  ) {}

  async get(clientId: string): Promise<Client | undefined> {
    const key = `${this.ssmKeyPrefix}/${clientId}`;

    const cached = this.cache.get<Client>(key);

    if (cached) return cached;

    let client: string | undefined;

    try {
      const response = await this.ssmClient.send(
        new GetParameterCommand({
          Name: key,
        })
      );
      client = response.Parameter?.Value;
    } catch (error) {
      this.logger.error('failed to obtain client configuration', {
        error,
        clientId,
        key,
      });
      return;
    }

    const { data, error, success } = $ClientProcessed.safeParse(client);

    if (!success) {
      this.logger.error('Failed to parse client configuration', error, {
        clientId,
      });
      return;
    }

    this.cache.set<Client>(key, data);

    return data;
  }
}
