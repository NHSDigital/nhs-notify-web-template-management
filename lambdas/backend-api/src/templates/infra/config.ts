import z from 'zod';

const $Env = z.object({
  DEFAULT_LETTER_SUPPLIER: z.string(),
  ENABLE_PROOFING: z.enum(['true', 'false']).default('false'),
  ENVIRONMENT: z.string(),
  REQUEST_PROOF_QUEUE_URL: z.string(),
  TEMPLATES_INTERNAL_BUCKET_NAME: z.string(),
  TEMPLATES_QUARANTINE_BUCKET_NAME: z.string(),
  TEMPLATES_DOWNLOAD_BUCKET_NAME: z.string(),
  TEMPLATES_TABLE_NAME: z.string(),
});

export function loadConfig() {
  const env = $Env.parse(process.env);

  return {
    defaultLetterSupplier: env.DEFAULT_LETTER_SUPPLIER,
    enableProofing: env.ENABLE_PROOFING === 'true',
    environment: env.ENVIRONMENT,
    internalBucket: env.TEMPLATES_INTERNAL_BUCKET_NAME,
    quarantineBucket: env.TEMPLATES_QUARANTINE_BUCKET_NAME,
    downloadBucket: env.TEMPLATES_DOWNLOAD_BUCKET_NAME,
    requestProofQueueUrl: env.REQUEST_PROOF_QUEUE_URL,
    templatesTableName: env.TEMPLATES_TABLE_NAME,
  };
}
