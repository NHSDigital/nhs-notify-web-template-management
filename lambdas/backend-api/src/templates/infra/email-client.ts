import { createMimeMessage, MailboxAddrObject } from 'mimetext';
import { SESClient, SendRawEmailCommand } from '@aws-sdk/client-ses';
import { TemplateDto } from 'nhs-notify-backend-client';
import { Logger } from 'nhs-notify-web-template-management-utils/logger';
import { LetterTemplate } from 'nhs-notify-web-template-management-utils';
import Handlebars from 'handlebars';
import { readFileSync } from 'node:fs';

export class EmailClient {
  constructor(
    private readonly sesClient: SESClient,
    private readonly senderEmail: string,
    private readonly recipientEmails: Record<string, string[]>,
    private readonly logger: Logger
  ) {}

  private getTemplateSubmittedEmailContent(
    { updatedAt, id, name }: LetterTemplate,
    supplier: string,
    proofFilenames: string[]
  ) {
    const emailTemplateContent = readFileSync(
      './email-template.html'
    ).toString();
    const htmlTemplate = Handlebars.compile(emailTemplateContent);

    return htmlTemplate({
      proofFilenames,
      supplier,
      timestamp: updatedAt,
      templateId: id,
      templateName: name,
    });
  }

  private async sendTemplateSubmittedEmailToSupplier(
    template: LetterTemplate,
    supplier: string,
    proofFilenames: string[]
  ) {
    const recipientEmailsForSupplier = this.recipientEmails[supplier];

    if ((recipientEmailsForSupplier?.length ?? 0) === 0) {
      this.logger.info({
        description:
          'Not sending template submitted email to supplier because no recipients are configured',
        templateId: template.id,
        supplier,
      });

      return;
    }

    const msg = createMimeMessage();
    msg.setSender({ name: 'NHS Notify', addr: this.senderEmail });

    const recipients: MailboxAddrObject[] = recipientEmailsForSupplier.map(
      (emailAddress) => ({ addr: emailAddress, type: 'Bcc' })
    );
    msg.setTo(recipients, { type: 'Bcc' });
    msg.setSubject(`${supplier} - Letter proof approved by an NHS Notify user`);
    msg.addMessage({
      contentType: 'text/html',
      data: this.getTemplateSubmittedEmailContent(
        template,
        supplier,
        proofFilenames
      ),
    });

    const command = new SendRawEmailCommand({
      RawMessage: {
        Data: Buffer.from(msg.asRaw()),
      },
    });

    const res = await this.sesClient.send(command);
    this.logger.info(res);
  }

  async sendTemplateSubmittedEmailToSuppliers(template: TemplateDto) {
    // cannot send if no email address provided
    if (!this.senderEmail) {
      this.logger.info({
        description:
          'Not sending template submitted email to suppliers because no email address is provided',
        templateId: template.id,
      });

      return;
    }

    // nothing to send if template is not a letter
    if (template.templateType !== 'LETTER') {
      this.logger.info({
        description:
          'Not sending template submitted email to suppliers because templateType is not LETTER',
        templateId: template.id,
      });

      return;
    }

    const proofsBySupplier: Record<string, string[]> = {};

    for (const { supplier, fileName } of Object.values(
      template.files.proofs ?? {}
    )) {
      const proofFilenames = [...(proofsBySupplier[supplier] ?? []), fileName];
      proofsBySupplier[supplier] = proofFilenames;
    }

    for (const [supplier, proofFilenames] of Object.entries(proofsBySupplier)) {
      await this.sendTemplateSubmittedEmailToSupplier(
        template,
        supplier,
        proofFilenames
      );
    }
  }
}
