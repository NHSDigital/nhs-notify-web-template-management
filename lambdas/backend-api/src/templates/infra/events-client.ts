import type { EventBridgeEvent } from 'aws-lambda';
import {
  EventBridgeClient,
  PutEventsCommand,
} from '@aws-sdk/client-eventbridge';
import { logger } from 'nhs-notify-web-template-management-utils/logger';

export class EventsClient {
  private client = new EventBridgeClient();

  constructor(
    private source: string,
    private busName: string = 'default'
  ) {}

  async putEvent<TDetailType extends string, TDetail>(
    event: Pick<
      EventBridgeEvent<TDetailType, TDetail>,
      'detail' | 'detail-type'
    >
  ): Promise<void> {
    logger.info('event', event);
    const { FailedEntryCount = 0, Entries = [] } = await this.client.send(
      new PutEventsCommand({
        Entries: [
          {
            Detail: JSON.stringify(event.detail),
            DetailType: event['detail-type'],
            EventBusName: this.busName,
            Source: this.source,
          },
        ],
      })
    );

    logger.info('event sent', { FailedEntryCount, Entries });

    if ((FailedEntryCount || 0) > 0) {
      throw new Error('Failed to send event to EventBridge.', {
        cause: Entries,
      });
    }
  }
}
