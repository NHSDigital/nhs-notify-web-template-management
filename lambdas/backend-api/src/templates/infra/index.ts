export * from './events-client';
export * from './letter-upload-repository';
export * from './template';
export * from './template-repository';
