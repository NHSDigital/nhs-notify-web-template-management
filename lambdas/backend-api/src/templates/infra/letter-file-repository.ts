import {
  CopyObjectCommand,
  DeleteObjectCommand,
  S3Client,
} from '@aws-sdk/client-s3';

export class LetterFileRepository {
  protected readonly client = new S3Client();

  constructor(
    protected readonly quarantineBucketName: string,
    protected readonly internalBucketName: string
  ) {}

  async copyFromQuarantineToInternal(
    key: string,
    versionId: string,
    destinationKey?: string
  ) {
    await this.client.send(
      new CopyObjectCommand({
        CopySource: `/${this.quarantineBucketName}/${key}?versionId=${versionId}`,
        Bucket: this.internalBucketName,
        Key: destinationKey ?? key,
        MetadataDirective: 'COPY',
        TaggingDirective: 'COPY',
      })
    );
  }

  async deleteFromQuarantine(key: string, versionId: string) {
    await this.client.send(
      new DeleteObjectCommand({
        Bucket: this.quarantineBucketName,
        Key: key,
        VersionId: versionId,
      })
    );
  }
}
