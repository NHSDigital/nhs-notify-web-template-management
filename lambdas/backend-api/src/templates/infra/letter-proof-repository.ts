import { LetterFileRepository } from './letter-file-repository';

export type LetterProofMetadata = {
  templateId: string;
  fileName: string;
  supplier: string;
};

export class LetterProofRepository extends LetterFileRepository {
  static parseQuarantineKey(key: string): LetterProofMetadata {
    const keyParts = key.split('/');
    const [fileType, supplier, templateId, fileName] = keyParts;
    const extension = fileName.split('.').at(-1);

    if (
      keyParts.length !== 4 ||
      fileType !== 'proofs' ||
      extension?.toLowerCase() !== 'pdf'
    ) {
      throw new Error(`Unexpected object key "${key}"`);
    }

    return {
      templateId,
      fileName,
      supplier,
    };
  }

  static getInternalKey(owner: string, templateId: string, fileName: string) {
    return `proofs/${owner}/${templateId}/${fileName}`;
  }

  static getDownloadKey(owner: string, templateId: string, fileName: string) {
    return `${owner}/proofs/${templateId}/${fileName}`;
  }
}
