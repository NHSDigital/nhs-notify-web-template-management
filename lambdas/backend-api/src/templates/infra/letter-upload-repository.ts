import { z } from 'zod/v4';
import { ErrorCase } from 'nhs-notify-backend-client';
import type { FileType, User } from 'nhs-notify-web-template-management-utils';
import {
  GetObjectCommand,
  NotFound,
  PutObjectCommand,
} from '@aws-sdk/client-s3';
import { ApplicationResult, failure, success } from '../../utils';
import { LetterFileRepository } from './letter-file-repository';

export type LetterUploadMetadata = {
  'file-type': FileType;
  'client-id': string;
  'template-id': string;
  'version-id': string;
};

const $FileType: z.ZodType<FileType> = z.enum([
  'pdf-template',
  'test-data',
  'proofs',
]);

const $LetterUploadMetadata: z.ZodType<LetterUploadMetadata> = z.object({
  'file-type': $FileType,
  'client-id': z.string(),
  'template-id': z.string(),
  'version-id': z.string(),
});

export class LetterUploadRepository extends LetterFileRepository {
  async upload(
    templateId: string,
    user: User,
    versionId: string,
    pdf: File,
    csv?: File
  ): Promise<ApplicationResult<null>> {
    const pdfKey = this.key(
      'pdf-template',
      user.clientId,
      templateId,
      versionId
    );

    const commands: PutObjectCommand[] = [
      new PutObjectCommand({
        Bucket: this.quarantineBucketName,
        Key: pdfKey,
        Body: await pdf.bytes(),
        ChecksumAlgorithm: 'SHA256',
        Metadata: LetterUploadRepository.metadata(
          user.clientId,
          templateId,
          versionId,
          'pdf-template'
        ),
      }),
    ];

    if (csv) {
      const csvKey = this.key(
        'test-data',
        user.clientId,
        templateId,
        versionId
      );

      commands.push(
        new PutObjectCommand({
          Bucket: this.quarantineBucketName,
          Key: csvKey,
          Body: await csv.bytes(),
          ChecksumAlgorithm: 'SHA256',
          Metadata: LetterUploadRepository.metadata(
            user.clientId,
            templateId,
            versionId,
            'test-data'
          ),
        })
      );
    }

    try {
      await Promise.all(commands.map((cmd) => this.client.send(cmd)));
      return success(null);
    } catch (error) {
      return failure(
        ErrorCase.INTERNAL,
        'Failed to upload letter files',
        error
      );
    }
  }

  async download(
    templateId: string,
    owner: string,
    fileType: FileType,
    versionId: string
  ): Promise<Uint8Array | void> {
    try {
      const { Body } = await this.client.send(
        new GetObjectCommand({
          Bucket: this.internalBucketName,
          Key: this.key(fileType, owner, templateId, versionId),
        })
      );

      return Body?.transformToByteArray();
    } catch (error) {
      if (error instanceof NotFound) {
        return;
      }

      throw error;
    }
  }

  static parseKey(key: string): LetterUploadMetadata {
    const keyParts = key.split('/');
    const [type, clientId, templateId, filename = ''] = keyParts;
    const filenameParts = filename.split('.');
    const [versionId, extension] = filenameParts;

    if (keyParts.length !== 4 || filenameParts.length !== 2) {
      throw new Error(`Unexpected object key "${key}"`);
    }

    const parsed = LetterUploadRepository.metadata(
      clientId,
      templateId,
      versionId,
      type
    );

    const expectedExtension =
      parsed['file-type'] === 'test-data' ? 'csv' : 'pdf';

    if (extension.toLowerCase() !== expectedExtension) {
      throw new Error(`Unexpected object key "${key}"`);
    }

    return parsed;
  }

  private key(
    type: FileType,
    clientId: string,
    templateId: string,
    versionId: string
  ) {
    return `${type}/${clientId}/${templateId}/${versionId}.${type === 'pdf-template' ? 'pdf' : 'csv'}`;
  }

  private static metadata(
    clientId: string,
    templateId: string,
    versionId: string,
    type: string
  ): LetterUploadMetadata {
    return $LetterUploadMetadata.parse({
      'client-id': clientId,
      'file-type': type,
      'template-id': templateId,
      'version-id': versionId,
    });
  }
}
