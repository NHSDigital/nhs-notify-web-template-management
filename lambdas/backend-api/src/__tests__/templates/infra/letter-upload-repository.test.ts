import {
  CopyObjectCommand,
  DeleteObjectCommand,
  GetObjectCommand,
  GetObjectCommandOutput,
  NotFound,
  PutObjectCommand,
  S3Client,
} from '@aws-sdk/client-s3';
import 'aws-sdk-client-mock-jest';
import { mockClient } from 'aws-sdk-client-mock';
import { LetterUploadRepository } from '../../../templates/infra/letter-upload-repository';
import { mock } from 'jest-mock-extended';

const quarantineBucketName = 'quarantine-bucket';
const internalBucketName = 'internal-bucket';

const setup = () => {
  const s3Client = mockClient(S3Client);

  const letterUploadRepository = new LetterUploadRepository(
    quarantineBucketName,
    internalBucketName
  );

  return { letterUploadRepository, mocks: { s3Client } };
};

describe('LetterUploadRepository', () => {
  const templateId = 'B9B7756DDDA9';
  const owner = '3A1F94D78582';
  const versionId = 'A6C177531604';

  const pdfBytes = new Blob(['pdf_data']);
  const csvBytes = new Blob(['csv_data']);
  const pdfFilename = 'template.pdf';
  const csvFilename = 'test-data.csv';

  describe('upload', () => {
    const pdf = new File([pdfBytes], pdfFilename, {
      type: 'application/pdf',
    });
    const csv = new File([csvBytes], csvFilename, {
      type: 'text/csv',
    });

    test('uploads both PDF template and test data CSV', async () => {
      const { letterUploadRepository, mocks } = setup();

      await letterUploadRepository.upload(
        templateId,
        owner,
        versionId,
        pdf,
        csv
      );

      expect(mocks.s3Client).toHaveReceivedCommandTimes(PutObjectCommand, 2);

      expect(mocks.s3Client).toHaveReceivedCommandWith(PutObjectCommand, {
        Bucket: quarantineBucketName,
        Key: `pdf-template/${owner}/${templateId}/${versionId}.pdf`,
        Body: new Uint8Array(await pdfBytes.arrayBuffer()),
        Metadata: {
          owner,
          'file-type': 'pdf-template',
          'template-id': templateId,
          'version-id': versionId,
        },
      });

      expect(mocks.s3Client).toHaveReceivedCommandWith(PutObjectCommand, {
        Bucket: quarantineBucketName,
        Key: `test-data/${owner}/${templateId}/${versionId}.csv`,
        Body: new Uint8Array(await csvBytes.arrayBuffer()),
        Metadata: {
          owner,
          'file-type': 'test-data',
          'template-id': templateId,
          'version-id': versionId,
        },
      });
    });

    test('uploads only the PDF template when test data CSV is not present', async () => {
      const { letterUploadRepository, mocks } = setup();
      await letterUploadRepository.upload(templateId, owner, versionId, pdf);

      expect(mocks.s3Client).toHaveReceivedCommandTimes(PutObjectCommand, 1);

      expect(mocks.s3Client).toHaveReceivedCommandWith(PutObjectCommand, {
        Bucket: quarantineBucketName,
        Key: `pdf-template/${owner}/${templateId}/${versionId}.pdf`,
        Body: new Uint8Array(await pdfBytes.arrayBuffer()),
        Metadata: {
          owner,
          'file-type': 'pdf-template',
          'template-id': templateId,
          'version-id': versionId,
        },
      });
    });

    test('returns error when upload fails', async () => {
      const { letterUploadRepository, mocks } = setup();

      const err = new Error('could not upload');

      mocks.s3Client.on(PutObjectCommand).rejects(err);

      const result = await letterUploadRepository.upload(
        templateId,
        owner,
        versionId,
        pdf,
        csv
      );

      expect(result).toEqual({
        error: {
          actualError: err,
          code: 500,
          details: undefined,
          message: 'Failed to upload letter files',
        },
      });

      expect(mocks.s3Client).toHaveReceivedCommandTimes(PutObjectCommand, 2);
    });
  });

  describe('copyFromQuarantineToInternal', () => {
    it('copies pdf template files from quarantine to internal', async () => {
      const { letterUploadRepository, mocks } = setup();

      await letterUploadRepository.copyFromQuarantineToInternal(
        's3/object/key',
        's3-object-version'
      );

      expect(mocks.s3Client).toHaveReceivedCommandWith(CopyObjectCommand, {
        Bucket: 'internal-bucket',
        CopySource:
          '/quarantine-bucket/s3/object/key?versionId=s3-object-version',
        Key: 's3/object/key',
        MetadataDirective: 'COPY',
        TaggingDirective: 'COPY',
      });
    });
  });

  describe('deleteFromQuarantine', () => {
    it('deletes files from quarantine', async () => {
      const { letterUploadRepository, mocks } = setup();

      await letterUploadRepository.deleteFromQuarantine('key', 'version');

      expect(mocks.s3Client).toHaveReceivedCommandWith(DeleteObjectCommand, {
        Bucket: 'quarantine-bucket',
        Key: 'key',
        VersionId: 'version',
      });
    });
  });

  describe('download', () => {
    it('gets the object from s3 and returns the body as Uint8Array', async () => {
      const { letterUploadRepository, mocks } = setup();

      const expected = Uint8Array.from('hello');
      mocks.s3Client.on(GetObjectCommand).resolves({
        Body: mock<GetObjectCommandOutput['Body']>({
          transformToByteArray: async () => expected,
        }),
      });

      const res = await letterUploadRepository.download(
        {
          id: 'template-id',
          owner: 'template-owner',
        },
        'pdf-template',
        'file-version-id'
      );

      expect(mocks.s3Client).toHaveReceivedCommandWith(GetObjectCommand, {
        Bucket: 'internal-bucket',
        Key: 'pdf-template/template-owner/template-id/file-version-id.pdf',
      });
      expect(res).toBe(expected);
    });

    it('returns void if command throws NotFound', async () => {
      const { letterUploadRepository, mocks } = setup();
      mocks.s3Client
        .on(GetObjectCommand)
        .rejects(new NotFound({ $metadata: {}, message: 'NotFound' }));

      const res = await letterUploadRepository.download(
        {
          id: 'template-id',
          owner: 'template-owner',
        },
        'pdf-template',
        'file-version-id'
      );

      expect(res).toBeUndefined();
    });

    it('raises other exceptions', async () => {
      const { letterUploadRepository, mocks } = setup();
      mocks.s3Client.on(GetObjectCommand).rejects(new Error('oh no'));

      await expect(
        letterUploadRepository.download(
          {
            id: 'template-id',
            owner: 'template-owner',
          },
          'pdf-template',
          'file-version-id'
        )
      ).rejects.toThrow('oh no');
    });
  });

  describe('static parseKey', () => {
    it('returns metadata from valid pdf key', () => {
      expect(
        LetterUploadRepository.parseKey(
          'pdf-template/owner-id/template-id/version-id.pdf'
        )
      ).toEqual({
        'file-type': 'pdf-template',
        owner: 'owner-id',
        'template-id': 'template-id',
        'version-id': 'version-id',
      });
    });

    it('returns metadata from valid csv key', () => {
      expect(
        LetterUploadRepository.parseKey(
          'test-data/owner-id/template-id/version-id.csv'
        )
      ).toEqual({
        'file-type': 'test-data',
        owner: 'owner-id',
        'template-id': 'template-id',
        'version-id': 'version-id',
      });
    });

    it('errors if key if too long', () => {
      expect(() =>
        LetterUploadRepository.parseKey(
          'test-data/owner-id/template-id/unexpected-path/version-id.csv'
        )
      ).toThrowErrorMatchingSnapshot();
    });

    it('errors if key if too short', () => {
      expect(() =>
        LetterUploadRepository.parseKey('test-data/owner-id/version-id.csv')
      ).toThrowErrorMatchingSnapshot();
    });

    it('errors if invalid file type segment', () => {
      expect(() =>
        LetterUploadRepository.parseKey(
          'unexpected/owner-id/template-id/version-id.csv'
        )
      ).toThrowErrorMatchingSnapshot();
    });

    it('errors if no file extension', () => {
      expect(() =>
        LetterUploadRepository.parseKey(
          'test-data/owner-id/template-id/version-id'
        )
      ).toThrowErrorMatchingSnapshot();
    });

    it('errors if filename has too many parts', () => {
      expect(() =>
        LetterUploadRepository.parseKey(
          'test-data/owner-id/template-id/version-id.unexpected.csv'
        )
      ).toThrowErrorMatchingSnapshot();
    });

    it('errors if file extension does not match csv file type', () => {
      expect(() =>
        LetterUploadRepository.parseKey(
          'test-data/owner-id/template-id/version-id.pdf'
        )
      ).toThrowErrorMatchingSnapshot();
    });

    it('errors if file extension does not match pdf file type', () => {
      expect(() =>
        LetterUploadRepository.parseKey(
          'pdf-template/owner-id/template-id/version-id.csv'
        )
      ).toThrowErrorMatchingSnapshot();
    });
  });
});
