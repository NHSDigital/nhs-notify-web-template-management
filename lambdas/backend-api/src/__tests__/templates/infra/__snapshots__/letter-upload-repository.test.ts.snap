// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LetterUploadRepository static parseKey errors if file extension does not match csv file type 1`] = `"Unexpected object key "test-data/owner-id/template-id/version-id.pdf""`;

exports[`LetterUploadRepository static parseKey errors if file extension does not match pdf file type 1`] = `"Unexpected object key "pdf-template/owner-id/template-id/version-id.csv""`;

exports[`LetterUploadRepository static parseKey errors if filename has too many parts 1`] = `"Unexpected object key "test-data/owner-id/template-id/version-id.unexpected.csv""`;

exports[`LetterUploadRepository static parseKey errors if invalid file type segment 1`] = `
"[
  {
    "received": "unexpected",
    "code": "invalid_enum_value",
    "options": [
      "pdf-template",
      "test-data",
      "proofs"
    ],
    "path": [
      "file-type"
    ],
    "message": "Invalid enum value. Expected 'pdf-template' | 'test-data' | 'proofs', received 'unexpected'"
  }
]"
`;

exports[`LetterUploadRepository static parseKey errors if key if too long 1`] = `"Unexpected object key "test-data/owner-id/template-id/unexpected-path/version-id.csv""`;

exports[`LetterUploadRepository static parseKey errors if key if too short 1`] = `"Unexpected object key "test-data/owner-id/version-id.csv""`;

exports[`LetterUploadRepository static parseKey errors if no file extension 1`] = `"Unexpected object key "test-data/owner-id/template-id/version-id""`;
